substrate-node/tests/TfChainClient.py

```diff
@@ -0,0 +1,841 @@
+        logging.info("Setting up %s (%s)", name,
+                     self._predefined_keys[name].ss58_address)
+
+        substrate = self._connect_to_server(f"ws://127.0.0.1:{port}")
+
+        insert_key_params = [
+            "tft!", f"//{name}", self._predefined_keys[name].public_key.hex()]
+        substrate.rpc_request("author_insertKey", insert_key_params)
+
+        insert_key_params = [
+            "smct", f"//{name}", self._predefined_keys[name].public_key.hex()]
+        substrate.rpc_request("author_insertKey", insert_key_params)
+
+        if create_twin:
+            self.user_accept_tc(port=port, who=name)
+            self.create_twin(port=port, who=name)
+
+    def _check_signer(self, who: str):
+        _who = who.title()
+        assert _who in self._predefined_keys, f"{who} is not a predefined account, use one of {self._predefined_keys.keys()}"
+        return _who
+
+    def _check_events(self, events: list = [], expected_events: list = []):
+        logging.info("Events: %s", json.dumps(events))
+
+        # This was a sudo call that failed
+        for event in events:
+            if event["event_id"] == "Sudid" and "Err" in event["attributes"]:
+                raise Exception(event["attributes"])
+
+        for expected_event in expected_events:
+            check = False
+            for event in events:
+                check = all(item in event.keys(
+                ) and event[item] == expected_event[item] for item in expected_event.keys())
+                if check:
+                    logging.info("Found event %s", expected_event)
+                    break
+            if not check:
+                raise Exception(
+                    f"Expected the event {expected_event} in {events}, no match found!")
+
+    def _sign_extrinsic_submit_check_response(self, substrate, call, who: str, expected_events: list = []):
+        _who = who.title()
+        if _who == "Sudo":
+            call = substrate.compose_call("Sudo", "sudo", {
+                "call": call
+            })
+            _who = "Alice"
+        else:
+            assert _who in self._predefined_keys, f"{who} is not a predefined account, use one of {self._predefined_keys.keys()}"
+
+        logging.info("Sending signed transaction: %s", call)
+        signed_call = substrate.create_signed_extrinsic(
+            call, self._predefined_keys[_who])
+
+        response = substrate.submit_extrinsic(
+            signed_call, wait_for_finalization=False, wait_for_inclusion=True)
+        logging.info("Reponse is %s", response)
+        if response.error_message:
+            raise Exception(response.error_message)
+
+        self._check_events([event.value["event"]
+                           for event in response.triggered_events], expected_events)
+
+    def setup_alice(self, create_twin: bool = False, port: int = DEFAULT_PORT):
+        self._setup_predefined_account(
```

Comment:
Okay, I see what you are trying to do here. We don't need to insert the keys in the keystore for `tft!` and `smct` for every well known key.

You only insert 1 key per offchain worker, for `tft!` it could be Alice and for `smct` it could be Bob

Reply:
What happens if you do do that?

Reply:
in this case you populate the keystore with all well known keys. The offchain worker will fire and take the first available key in the keystore. While this is fine, in production we wont do this.